src/ucp/core/ucp_context.c

```diff
@@ -438,14 +487,127 @@ void ucp_config_release(ucp_config_t *config)
 ucs_status_t ucp_config_modify(ucp_config_t *config, const char *name,
                                const char *value)
 {
-    return ucs_config_parser_set_value(config, ucp_config_table, name, value);
+    ucs_config_cached_key_t *key_val;
+    ucs_status_t ucs_status;
+
+    ucs_status = ucs_config_parser_set_value(config, ucp_config_table, name,
+                                             value);
+    if (ucs_status == UCS_OK) {
+        return UCS_OK;
+    }
+
+    ucs_status = ucp_cached_key_alloc(&key_val, name, value);
+    if (ucs_status == UCS_OK) {
+        ucs_list_add_tail(&config->list, &key_val->list);
+    }
+
+    return UCS_ERR_NO_ELEM;
+}
+
+static
+ucs_status_t uct_cfg_build(uct_config_bundle_t **bundles, unsigned int *idx_cfg,
+                           ucs_config_global_list_entry_t *entry)
+{
+    uct_config_bundle_t *bundle;
+    ucs_status_t status;
+
+    if (entry->size == 0) {
+        return UCS_ERR_REJECTED;
+    }
+
+    status = uct_config_read(&bundle, entry->table, entry->size, NULL,
+                             entry->prefix);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    bundles[*idx_cfg] = bundle;
+    *idx_cfg          = *idx_cfg + 1;
+    return UCS_OK;
 }
 
 void ucp_config_print(const ucp_config_t *config, FILE *stream,
                       const char *title, ucs_config_print_flags_t print_flags)
 {
+    unsigned int idx_bundle = 0;
+    uct_config_bundle_t **bundles;
+    unsigned int idx_cmpt, num_cmpts;
+    uct_component_h *cmpts;
+    ucs_config_global_list_entry_t *entry;
+    uct_tl_t *tl;
+    ucs_config_cached_key_t *key_val;
+    unsigned int cfg_id;
+    ucs_status_t status;
+
     ucs_config_parser_print_opts(stream, title, config, ucp_config_table,
                                  NULL, UCS_DEFAULT_ENV_PREFIX, print_flags);
+
+    status = uct_query_components(&cmpts, &num_cmpts);
+    if (status != UCS_OK) {
+        ucs_error("failed to load UCT modules");
+        return;
+    }
+
+    bundles = ucs_calloc(num_cmpts * 3 + 1, sizeof(*bundles), "uct cfg bundle");
+    if (bundles == NULL) {
+        ucs_error("No memory, can't check UCT configurations");
+        return;
```

Comment:
why do we need an array of bundles? can't we just print the list of cached key/value pairs?

Reply:
The cached key/value pairs are in below format.
https://github.com/openucx/ucx/search?q=uct_config_modify+filename%3Atest_rc.cc+path%3A%2Ftest%2Fgtest%2Fuct%2Fib&type=code
However, the printed format are UCX_XXX(full path), so we need construct the bundle and apply the key/value paris, then print them out.

Reply:
no need to print full path
can just print whatever user has passed as key

Reply:
@yosefe In this way, the user maybe not easy to know whether the UCT key/val pair are really being applied since the ucp_config_print won't print the UCT configurations in the similar format as ```ucx_info -c```. If some of the key/val are not effective, they'll be also printed, which seems strange.

Reply:
I think it's not so important: if we really want to check the key/value pair, should do it in ucp_config_modify() - or no do it at all.
In any case, if we do move it to ucp_config_read(), need to refactor the code to simplify it. for example, no need to allocate an array of bundles. can check one at a time.

Reply:
If there's no problem when implementing checking key/value in ```ucp_config_modify```, then I'll simplify the code here i.e. print key/value directly.

Reply:
Done